libc-top-half/musl/src/unistd/chdir.c

```diff
@@ -1,7 +1,46 @@
 #include <unistd.h>
 #include "syscall.h"
 
+#ifndef __wasilibc_unmodified_upstream
+#include <string.h>
+#include <errno.h>
+#include <limits.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <wasi/libc.h>
+#include <wasi/libc-find-relpath.h>
+#endif
+
+extern char __wasilibc_cwd[PATH_MAX];
+
 int chdir(const char *path)
 {
+#ifdef __wasilibc_unmodified_upstream // WASI libc uses a custom chdir
 	return syscall(SYS_chdir, path);
+#else
+        // Find a preopen'd directory as well as a relative path we're anchored
+        // from which we're changing directories to.
+        char *relative;
+        const char *abs;
+        int parent_fd = __wasilibc_find_relpath(path, &abs, &relative);
+        if (parent_fd == -1) {
+                errno = ENOENT;
+                return -1;
+        }
+
+        // Make sure that this directory we're accessing is indeed a directory.
+        int fd = __wasilibc_openat_nomode(parent_fd, relative, O_DIRECTORY | O_RDONLY);
+        if (fd == -1) {
+                free(relative);
+                return -1;
+        }
+
+        // Copy over our new absolute path into `__wasilibc_cwd`
+        close(fd);
+        strcpy(__wasilibc_cwd, abs);
+        if (strcmp(relative, ".") != 0)
+            strcpy(__wasilibc_cwd + strlen(abs), relative);
+        free(relative);
+        return 0;
+#endif
```

Comment:
In cases like this where the entire musl function body is replaced with WASI-specific code, it's better to just create a new file rather than making musl's conditional. We can put this in libc-bottom-half/sources/chdir.c.